presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/MetadataDeleteOptimizer.java

```diff
@@ -78,11 +78,11 @@ private Optimizer(Session session, Metadata metadata, PlanNodeIdAllocator idAllo
         public PlanNode visitTableCommit(TableCommitNode node, RewriteContext<Void> context)
         {
             Optional<DeleteNode> delete = findNode(node.getSource(), DeleteNode.class);
-            if (!delete.isPresent() || !metadata.supportsMetadataDelete(session, delete.get().getTarget().getHandle())) {
+            if (!delete.isPresent()) {
                 return context.defaultRewrite(node);
             }
             Optional<TableScanNode> tableScan = findNode(delete.get().getSource(), TableScanNode.class);
-            if (!tableScan.isPresent()) {
+            if (!tableScan.isPresent() || !metadata.supportsMetadataDelete(session, tableScan.get().getTable(), tableScan.get().getLayout().get())) {
```

Comment:
This might be easier to read if you split it up:
```java
if (!tableScan.isPresent()) {
    return context.defaultRewrite(node);
}
TableScanNode scan = tableScan.get();
if (!metadata.supportsMetadataDelete(session, scan.getTable(), scan.getLayout().get())) {
    return context.defaultRewrite(node);
}